File: android/app/src/main/java/com/lilru/liftr/data/CoinManager.kt
package com.lilru.liftr.data

import androidx.compose.ui.graphics.Color
import com.lilru.liftr.ui.AppSnackbar
import io.github.jan.supabase.SupabaseClient
import io.github.jan.supabase.auth.auth
import io.github.jan.supabase.postgrest.from
import io.github.jan.supabase.postgrest.postgrest
import io.github.jan.supabase.postgrest.query.Columns
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.asSharedFlow
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.put
import org.json.JSONArray

object CoinManager {
    @Volatile
    var balance: Int = 0
        private set

    private var lastEarnToastAtMs: Long = 0L
    private var hasEstablishedBalanceBaseline = false
    private val mutex = Mutex()
    private const val EARN_TOAST_DEBOUNCE_MS = 2_000L
    private const val EARN_TOAST_MAX_AGE_MS = 120_000L

    private val _transactionHistoryRefresh = MutableSharedFlow<Unit>(extraBufferCapacity = 1)
    val transactionHistoryRefresh = _transactionHistoryRefresh.asSharedFlow()

    fun notifyTransactionHistoryShouldRefresh() {
        _transactionHistoryRefresh.tryEmit(Unit)
    }

    fun syncBalance(value: Int) {
        balance = maxOf(0, value)
        hasEstablishedBalanceBaseline = true
    }

    fun resetSession() {
        balance = 0
        hasEstablishedBalanceBaseline = false
        lastEarnToastAtMs = 0L
    }

    suspend fun refreshBalance(
        supabase: SupabaseClient,
        notifyIfEarned: Boolean = false
    ) {
        val me = supabase.auth.currentUserOrNull()?.id ?: return
        val previousBalance = balance
        runCatching {
            @Serializable
            data class Row(@SerialName("coins_balance") val coinsBalance: Int? = null)
            val res = supabase.from(BackendContracts.Tables.PROFILES)
                .select(columns = Columns.raw(BackendContracts.ProfileColumns.COINS_BALANCE)) {
                    filter { eq("user_id", me) }
                    limit(1)
                }
            val next = maxOf(0, SupabaseResponseDecoding.decodeListOrObject<Row>(res.data).firstOrNull()?.coinsBalance ?: 0)
            balance = next
            if (!hasEstablishedBalanceBaseline) {
                hasEstablishedBalanceBaseline = true
                return@runCatching
            }
            if (notifyIfEarned && next > previousBalance) {
                notifyTransactionHistoryShouldRefresh()
                showEarnToast(supabase, next - previousBalance)
            }
        }
    }

    suspend fun refreshBalanceAfterMutation(
        supabase: SupabaseClient,
        notifyIfEarned: Boolean = true
    ) {
        refreshBalance(supabase, notifyIfEarned)
    }

    private suspend fun showEarnToast(supabase: SupabaseClient, fallbackDelta: Int) {
        val now = System.currentTimeMillis()
        mutex.withLock {
            if (now - lastEarnToastAtMs < EARN_TOAST_DEBOUNCE_MS) return
        }

        val message = runCatching {
            val res = supabase.postgrest.rpc(
                BackendContracts.Rpc.LIST_MY_COIN_TRANSACTIONS_V1,
                buildJsonObject { put("p_limit", 5) }
            ) { }
            val arr = parseRpcArray(res.data)
            val now = System.currentTimeMillis()
            val recent = (0 until arr.length()).mapNotNull { idx ->
                arr.optJSONObject(idx)
            }.firstOrNull { o ->
                o.optInt("amount") > 0 && isRecentEarn(o.optString("created_at"), now)
            } ?: return
            val amount = recent.optInt("amount")
            val label = displayLabel(recent.optString("action_type"))
            "+$amount Liftr Coins · $label"
        }.getOrNull() ?: return
        mutex.withLock { lastEarnToastAtMs = now }
        AppSnackbar.showSuccess(message)
    }

    private fun parseRpcArray(raw: String): JSONArray {
        val trimmed = raw.trim()
        return when {
            trimmed.startsWith("[") -> JSONArray(trimmed)
            trimmed.startsWith("{") -> JSONArray().put(org.json.JSONObject(trimmed))
            else -> JSONArray()
        }
    }

    private fun isRecentEarn(createdAt: String?, nowMs: Long): Boolean {
        if (createdAt.isNullOrBlank()) return false
        val createdMs = runCatching {
            java.time.Instant.parse(createdAt).toEpochMilli()
        }.getOrNull() ?: return false
        val age = nowMs - createdMs
        return age in 0..EARN_TOAST_MAX_AGE_MS
    }

    fun formattedAmount(amount: Int): String =
        if (amount >= 0) "+$amount" else "$amount"

    fun sourceKey(actionType: String): String = when (actionType) {
        "workout_logged",
        "workout_coin_doubling_v1",
        "workout_economy_rebalance_v1",
        "workout_economy_reduction_30pct_v1" -> "workouts"
        "workout_pet_training_bonus" -> "pet_workout_bonus"
        "pet_coins_generated",
        "pet_passive_economy_rebalance_v1",
        "pet_passive_economy_reduction_30pct_v1" -> "pet_coins"
        "like_given", "comment_added", "user_followed", "earned_follower" -> "social"
        "nutrition_ingredient_logged", "nutrition_recipe_logged",
        "nutrition_ingredient_created", "nutrition_recipe_created" -> "nutrition"
        "achievement_unlocked", "achievement_unlocked_bronze",
        "achievement_unlocked_silver", "achievement_unlocked_gold" -> "achievements"
        "weekly_goal_perfect_week", "workout_consistency_streak" -> "goals_streaks"
        "competition_bet_win", "competition_bet_refund_draw",
        "competition_bet_refund_cancelled", "competition_bet_escrow" -> "competition"
        "pet_combat_reward" -> "pet_combat"
        else -> "other"
    }

    fun displayLabel(actionType: String): String = when (actionType) {
        "like_given" -> "Like given"
        "comment_added" -> "Comment added"
        "user_followed" -> "Follow"
        "earned_follower" -> "New follower"
        "achievement_unlocked" -> "Achievement"
        "workout_logged" -> "Workout logged"
        "workout_pet_training_bonus" -> "Pet workout bonus"
        "workout_coin_doubling_v1" -> "Workout reward boost"
        "workout_economy_rebalance_v1" -> "Workout economy update"
        "pet_passive_economy_rebalance_v1" -> "Pet passive rebalance"
        "workout_economy_reduction_30pct_v1" -> "Workout economy adjustment"
        "pet_passive_economy_reduction_30pct_v1" -> "Pet passive adjustment"
        "weekly_goal_perfect_week" -> "Perfect week"
        "workout_consistency_streak" -> "7-day streak"
        "competition_bet_escrow" -> "Competition stake"
        "competition_bet_win" -> "Competition win"
        "competition_bet_refund_draw" -> "Competition draw refund"
        "competition_bet_refund_cancelled" -> "Competition refund"
        "pet_coins_generated" -> "Pet coins"
        "pet_market_purchase" -> "Pet shop purchase"
        "pet_egg_reroll" -> "Egg reroll"
        "pet_rarity_upgrade" -> "Pet rarity upgrade"
        "nutrition_ingredient_logged" -> "Food logged"
        "nutrition_recipe_logged" -> "Recipe logged"
        "nutrition_ingredient_created" -> "Ingredient created"
        "nutrition_recipe_created" -> "Recipe created"
        "" -> "Coins earned"
        else -> actionType.replace('_', ' ').replaceFirstChar { it.uppercase() }
    }

    fun sourceCategoryLabel(key: String): String = when (key) {
        "workouts" -> "Workouts"
        "pet_workout_bonus" -> "Pet workout bonus"
        "pet_coins" -> "Pet coins"
        "social" -> "Social"
        "nutrition" -> "Nutrition"
        "achievements" -> "Achievements"
        "goals_streaks" -> "Goals & streaks"
        "competition" -> "Competition"
        "pet_combat" -> "Pet combat"
        "other" -> "Other"
        else -> key.replace('_', ' ').replaceFirstChar { it.uppercase() }
    }

    fun sourceCategoryColor(key: String): Color = when (key) {
        "workouts" -> Color(0xFF1E88E5)
        "pet_workout_bonus" -> Color(0xFF9C27B0)
        "pet_coins" -> Color(0xFFFFC107)
        "social" -> Color(0xFFE91E63)
        "nutrition" -> Color(0xFF4CAF50)
        "achievements" -> Color(0xFFFF5722)
        "goals_streaks" -> Color(0xFF00BCD4)
        "competition" -> Color(0xFF795548)
        "pet_combat" -> Color(0xFF673AB7)
        else -> Color(0xFF9E9E9E)
    }
}

File: Liftr/CoinManager.swift
import Foundation
import SwiftUI
import Supabase

enum CoinSourcesPeriod: String, CaseIterable, Identifiable {
    case week
    case month
    case year
    case allTime

    var id: String { rawValue }

    var label: String {
        switch self {
        case .week: return "Week"
        case .month: return "Month"
        case .year: return "Year"
        case .allTime: return "All time"
        }
    }
}

struct CoinSourceSlice: Identifiable {
    let sourceKey: String
    let totalAmount: Int

    var id: String { sourceKey }
    var label: String { CoinManager.sourceCategoryLabel(for: sourceKey) }
    var color: Color { CoinManager.sourceCategoryColor(for: sourceKey) }
}

@MainActor
final class CoinManager: ObservableObject {
    static let shared = CoinManager()

    @Published private(set) var balance: Int = 0

    private struct BalanceRow: Decodable {
        let coins_balance: Int?
    }

    private struct TransactionRow: Decodable {
        let amount: Int
        let action_type: String
        let created_at: Date
    }

    private var lastEarnToastAt: Date?
    private let earnToastDebounce: TimeInterval = 2
    private var hasEstablishedBalanceBaseline = false
    private let earnToastMaxAge: TimeInterval = 120

    private init() {}

    func syncBalance(_ value: Int) {
        balance = max(0, value)
        hasEstablishedBalanceBaseline = true
    }

    func resetSession() {
        balance = 0
        hasEstablishedBalanceBaseline = false
        lastEarnToastAt = nil
    }

    func refreshBalance(notifyIfEarned: Bool = false) async {
        guard let userId = AppState.shared.userId else { return }

        let previousBalance = balance
        let client = SupabaseManager.shared.client

        do {
            let rows: [BalanceRow] = try await client
                .from("profiles")
                .select("coins_balance")
                .eq("user_id", value: userId.uuidString)
                .limit(1)
                .execute()
                .value

            guard let row = rows.first else { return }
            balance = max(0, row.coins_balance ?? 0)

            if !hasEstablishedBalanceBaseline {
                hasEstablishedBalanceBaseline = true
                return
            }

            if notifyIfEarned, balance > previousBalance {
                await showEarnToast(userId: userId, fallbackDelta: balance - previousBalance)
            }
        } catch {
            print("CoinManager refreshBalance error:", error.localizedDescription)
        }
    }

    func refreshBalanceAfterMutation(notifyIfEarned: Bool = true) async {
        await refreshBalance(notifyIfEarned: notifyIfEarned)
    }

    private func showEarnToast(userId: UUID, fallbackDelta: Int) async {
        let now = Date()
        if let last = lastEarnToastAt, now.timeIntervalSince(last) < earnToastDebounce {
            return
        }

        let client = SupabaseManager.shared.client

        do {
            var params: [String: AnyJSON] = [:]
            params["p_limit"] = AnyJSON(5)
            let res = try await client
                .rpc("list_my_coin_transactions_v1", params: params)
                .execute()
            let rows = try JSONDecoder.supabase().decode([TransactionRow].self, from: res.data)

            guard let recent = rows.first(where: { $0.amount > 0 }) else { return }
            let age = now.timeIntervalSince(recent.created_at)
            guard age >= 0, age < earnToastMaxAge else { return }

            let amount = recent.amount
            let label = Self.displayLabel(for: recent.action_type)
            let message = "+\(amount) Liftr Coins · \(label)"

            lastEarnToastAt = now
            AppState.shared.coinEarnToast = message
        } catch {
            return
        }
    }

    nonisolated static func formattedAmount(_ amount: Int) -> String {
        amount >= 0 ? "+\(amount)" : "\(amount)"
    }

    nonisolated static func displayLabel(for actionType: String) -> String {
        switch actionType {
        case "like_given": return "Like given"
        case "comment_added": return "Comment added"
        case "user_followed": return "Follow"
        case "earned_follower": return "New follower"
        case "achievement_unlocked": return "Achievement"
        case "workout_logged": return "Workout logged"
        case "workout_pet_training_bonus": return "Pet workout bonus"
        case "workout_coin_doubling_v1": return "Workout reward boost"
        case "workout_economy_rebalance_v1": return "Workout economy update"
        case "pet_passive_economy_rebalance_v1": return "Pet passive rebalance"
        case "workout_economy_reduction_30pct_v1": return "Workout economy adjustment"
        case "pet_passive_economy_reduction_30pct_v1": return "Pet passive adjustment"
        case "weekly_goal_perfect_week": return "Perfect week"
        case "workout_consistency_streak": return "7-day streak"
        case "competition_bet_escrow": return "Competition stake"
        case "competition_bet_win": return "Competition win"
        case "competition_bet_refund_draw": return "Competition draw refund"
        case "competition_bet_refund_cancelled": return "Competition refund"
        case "pet_coins_generated": return "Pet coins"
        case "pet_market_purchase": return "Pet shop purchase"
        case "pet_egg_reroll": return "Egg reroll"
        case "pet_rarity_upgrade": return "Pet rarity upgrade"
        case "nutrition_ingredient_logged": return "Food logged"
        case "nutrition_recipe_logged": return "Recipe logged"
        case "nutrition_ingredient_created": return "Ingredient created"
        case "nutrition_recipe_created": return "Recipe created"
        default:
            if actionType.isEmpty { return "Coins earned" }
            return actionType
                .replacingOccurrences(of: "_", with: " ")
                .capitalized
        }
    }

    nonisolated static func sourceCategoryLabel(for key: String) -> String {
        switch key {
        case "workouts": return "Workouts"
        case "pet_workout_bonus": return "Pet workout bonus"
        case "pet_coins": return "Pet coins"
        case "social": return "Social"
        case "nutrition": return "Nutrition"
        case "achievements": return "Achievements"
        case "goals_streaks": return "Goals & streaks"
        case "competition": return "Competition"
        case "pet_combat": return "Pet combat"
        case "other": return "Other"
        default:
            return key.replacingOccurrences(of: "_", with: " ").capitalized
        }
    }

    nonisolated static func sourceCategoryColor(for key: String) -> Color {
        switch key {
        case "workouts": return Color(red: 0.12, green: 0.53, blue: 0.90)
        case "pet_workout_bonus": return Color(red: 0.61, green: 0.15, blue: 0.69)
        case "pet_coins": return Color(red: 1.0, green: 0.76, blue: 0.03)
        case "social": return Color(red: 0.91, green: 0.12, blue: 0.39)
        case "nutrition": return Color(red: 0.30, green: 0.69, blue: 0.31)
        case "achievements": return Color(red: 1.0, green: 0.34, blue: 0.13)
        case "goals_streaks": return Color(red: 0.0, green: 0.74, blue: 0.83)
        case "competition": return Color(red: 0.47, green: 0.33, blue: 0.28)
        case "pet_combat": return Color(red: 0.40, green: 0.23, blue: 0.72)
        default: return Color(red: 0.62, green: 0.62, blue: 0.62)
        }
    }

    nonisolated static func coinSourcesTimeWindow(for period: CoinSourcesPeriod) -> (start: Date?, end: Date?) {
        guard period != .allTime else { return (nil, nil) }

        var cal = Calendar.current
        cal.timeZone = .current
        let now = Date()
        let dayStart = cal.startOfDay(for: now)

        switch period {
        case .allTime:
            return (nil, nil)
        case .week:
            let start = cal.date(byAdding: .day, value: -6, to: dayStart) ?? dayStart
            let end = cal.date(byAdding: .day, value: 1, to: dayStart) ?? now
            return (start, end)
        case .month:
            let start = cal.date(byAdding: .day, value: -29, to: dayStart) ?? dayStart
            let end = cal.date(byAdding: .day, value: 1, to: dayStart) ?? now
            return (start, end)
        case .year:
            let monthStart = cal.date(from: cal.dateComponents([.year, .month], from: now)) ?? now
            let start = cal.date(byAdding: .month, value: -11, to: monthStart) ?? monthStart
            let end = cal.date(byAdding: .day, value: 1, to: dayStart) ?? now
            return (start, end)
        }
    }
}
